RecoLocalMuon/GEMCSCSegment/plugins/GEMCSCSegFit.h

```diff
@@ -0,0 +1,179 @@
+#ifndef GEMCSCSegment_GEMCSCSegFit_h
+#define GEMCSCSegment_GEMCSCSegFit_h
+
+// GEMCSCSegFit.h extension to fit both CSCHits and GEMHits - Piet Verwilligen
+// Created: 21.04.2015
+
+/* Work with general Tracking Rechits, such that GEM and CSC rechits
+ * can be treated at the same level. In case needed the Tracking Rechit can be
+ * cast to the CSCRecHit2D or the GEMRecHit for access to more specific info.
+ */
+  
+// CSCSegFit.h - Segment fitting factored out of CSC segment builders - Tim Cox
+// Last mod: 03.02.2015
+
+/* This as an object which is initialized by a set of rechits (2 to 6) in a 
+ * specific CSC and has the functionality to make a least squares fit to a 
+ * straight line in 2-dim for those rechits.
+ * The covariance matrix and chi2 of the fit are calculated.
+ * The original code made use of CLHEP matrices but this version uses 
+ * ROOT SMatrices because they are  multithreading compatible.
+ * Because of this, the no. of rechits that can be handled is limited to
+ * a maximum of 6, one per layer of a CSC. This means maximum dimensions
+ * can be specified at compile time and hence satisfies SMatrix constraints.
+ * For 2 hits of course there is no fit - just draw a straight line between them.
+ * Details of the algorithm are in the .cc file
+ *
+ */
+   
+#include "DataFormats/TrackingRecHit/interface/TrackingRecHit.h"
+#include <DataFormats/CSCRecHit/interface/CSCRecHit2D.h>
+#include <DataFormats/GEMRecHit/interface/GEMRecHit.h>
+#include <DataFormats/MuonDetId/interface/CSCDetId.h>
+
+#include <Geometry/CSCGeometry/interface/CSCChamber.h>
+#include <Geometry/CSCGeometry/interface/CSCLayer.h>
+#include <Geometry/GEMGeometry/interface/GEMChamber.h>
+
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+
+#include <Math/Functions.h>
+#include <Math/SVector.h>
+#include <Math/SMatrix.h>
+
+#include <vector>
+
+class GEMCSCSegFit {
+
+public:
+
+// TYPES
+
+  // 16 x 16 Symmetric
+  typedef ROOT::Math::SMatrix<double,16,16,ROOT::Math::MatRepSym<double,16> > SMatrixSym16;
+
+  // 16 x 4
+  typedef ROOT::Math::SMatrix<double,16,4 > SMatrix16by4;
+
+  // 4 x 4 General + Symmetric
+  typedef ROOT::Math::SMatrix<double, 4 > SMatrix4;
+  typedef ROOT::Math::SMatrix<double,4,4,ROOT::Math::MatRepSym<double,4> > SMatrixSym4;
+
+  // 2 x 2 Symmetric
+  typedef ROOT::Math::SMatrix<double,2,2,ROOT::Math::MatRepSym<double,2> > SMatrixSym2;
+
+  // 4-dim vector
+  typedef ROOT::Math::SVector<double,4> SVector4;
+
+
+  // PUBLIC FUNCTIONS
+
+  //@@ WANT OBJECT TO CACHE THE SET OF HITS SO CANNOT PASS BY REF
+ GEMCSCSegFit(std::map<uint32_t, const CSCLayer*> csclayermap, std::map<uint32_t, const GEMEtaPartition*> gemrollmap, const std::vector<const TrackingRecHit*> hits) : 
```

Comment:
Shouldn't all these parameters be const references to avoid unnecessary copying?